wicketd/src/mgs/inventory.rs

```diff
@@ -0,0 +1,399 @@
+        // consumer is behind, we don't need to poll MGS until they can handle
+        // our results.
+        let (tx, rx) = mpsc::channel(1);
+
+        // "Poll immediately" also only needs a channel depth of 1: if there is
+        // already a message in this channel, we're already trying to poll ASAP.
+        let (poll_now_tx, poll_now_rx) = mpsc::channel(1);
+
+        let task = tokio::spawn(ignition_poller_task(
+            tx,
+            poll_now_rx,
+            mgs_client,
+            log,
+        ));
+
+        Self { task, rx, poll_now_tx }
+    }
+
+    /// Receive the next result from the ignition polling task.
+    pub(super) async fn recv(&mut self) -> PollIgnition {
+        // The task we spawned holds `tx` either until `rx` is dropped (which it
+        // obviously is not here, since we're using it!) or it panics, so we can
+        // unwrap here. The only way we panic is if our inner task already did.
+        self.rx.recv().await.expect("ignition polling task panicked")
+    }
+
+    pub(super) fn poll_now(&self) {
+        match self.poll_now_tx.try_send(()) {
+            // If we succeeded or there's already a "poll now" request sitting
+            // in the channel, we're done.
+            Ok(()) | Err(mpsc::error::TrySendError::Full(())) => (),
+            // If the channel is closed, that means our task has panicked -
+            // propogate that panic.
+            Err(mpsc::error::TrySendError::Closed(())) => {
+                panic!("ignition polling task panicked")
+            }
+        }
+    }
+}
+
+async fn ignition_poller_task(
+    tx: mpsc::Sender<PollIgnition>,
+    mut poll_now_rx: mpsc::Receiver<()>,
+    mgs_client: gateway_client::Client,
+    log: Logger,
+) {
+    let mut ticker = interval(POLL_FREQ_IGNITION);
+    ticker.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+
+    loop {
+        tokio::select! {
+            _ = ticker.tick() => (),
+            _ = poll_now_rx.recv() => {
+                ticker.reset();
+            }
+        }
+
+        let results = match mgs_client.ignition_list().await {
+            Ok(response) => response.into_inner(),
+            Err(err) => {
+                warn!(
+                    log, "Failed to poll MGS for ignition";
+                    "err" => %err,
+                );
+                continue;
+            }
+        };
+
+        let mut sps = BTreeMap::new();
+        let mgs_received = Instant::now();
+        for result in results {
+            sps.insert(result.id, result.details);
+        }
+
+        let emit = PollIgnition { sps, mgs_received };
+
+        // If our receiver is gone, we'll exit - there's no one left for
+        // us to send results to!
+        if tx.send(emit).await.is_err() {
+            warn!(log, "Receiver for ignition polling task is gone");
+            break;
+        }
+    }
+}
+
+pub(super) struct PollSp {
```

Comment:
Maybe call this something like `SpPollResponse` or `PolledSpData` instead?

Reply:
Renamed to `FetchedSpData` as part of the great poll -> fetch renaming.